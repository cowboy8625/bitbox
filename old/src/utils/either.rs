#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Either<L, R> {
    Left(L),
    Right(R),
}

impl<L, R> Either<L, R> {
    pub fn _map_left<F, T>(self, f: F) -> Either<T, R>
    where
        F: FnOnce(L) -> T,
    {
        match self {
            Either::Left(l) => Either::Left(f(l)),
            Either::Right(r) => Either::Right(r),
        }
    }

    pub fn _map_right<F, T>(self, f: F) -> Either<L, T>
    where
        F: FnOnce(R) -> T,
    {
        match self {
            Either::Left(l) => Either::Left(l),
            Either::Right(r) => Either::Right(f(r)),
        }
    }
}
