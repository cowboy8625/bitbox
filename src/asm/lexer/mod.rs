use std::{iter::Peekable, str::Chars};

pub fn lex(src: &str) -> Vec<Token> {
    Lexer::new(src).lex()
}

#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
pub struct Span {
    pub row_start: usize,
    pub row_end: usize,
    pub col_start: usize,
    pub col_end: usize,
    pub byte_start: usize,
    pub byte_end: usize,
}

impl From<(Span, Span)> for Span {
    fn from((start, end): (Span, Span)) -> Self {
        Self {
            row_start: start.row_start,
            row_end: end.row_end,
            col_start: start.col_start,
            col_end: end.col_end,
            byte_start: start.byte_start,
            byte_end: end.byte_end,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    KeywordPush,
    KeywordPop,
    KeywordLoad,
    KeywordAdd,
    KeywordSub,
    KeywordDiv,
    KeywordMul,
    KeywordEq,
    KeywordJne,
    KeywordInc,
    KeywordHult,
    KeywordPrintReg,
    KeywordAnd,
    KeywordOr,
    Number(u32),
    Identifier(String),
    Colon,
    Comma,
    Period,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    PercentSign,
    Ampersand,
    Equal,
    Delimiter,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

struct Lexer<'a> {
    src: Peekable<Chars<'a>>,
    span: Span,
    tokens: Vec<Token>,
    current: Option<char>,
    last_char: Option<char>,
}

impl<'a> Lexer<'a> {
    fn new(src: &'a str) -> Self {
        Self {
            src: src.chars().peekable(),
            span: Span::default(),
            tokens: Vec::new(),
            current: None,
            last_char: None,
        }
    }

    fn next(&mut self) -> Option<char> {
        self.last_char = self.current;
        self.current = self.src.next();
        match self.current {
            Some(c) if matches!(self.last_char, Some('\n')) => {
                self.span.row_end += 1;
                self.span.col_start = 0;
                self.span.col_end = 1;
                self.span.byte_end += 1;
                Some(c)
            }
            Some(c) => {
                self.span.col_end += 1;
                self.span.byte_end += 1;
                Some(c)
            }
            None => None,
        }
    }

    fn next_if(&mut self, pred: impl Fn(char) -> bool) -> Option<char> {
        match self.peek() {
            Some(c) if pred(c) => self.next(),
            _ => None,
        }
    }

    fn peek(&mut self) -> Option<char> {
        self.src.peek().copied()
    }

    fn add_token(&mut self, kind: TokenKind) {
        let span = self.span();
        self.tokens.push(Token { kind, span });
    }

    fn add_token_with_span(&mut self, kind: TokenKind, span: Span) {
        self.tokens.push(Token { kind, span });
    }

    fn span(&mut self) -> Span {
        let span = self.span;
        self.span = Span {
            row_start: span.row_end,
            col_start: span.col_end,
            byte_start: span.byte_end,
            ..span
        };
        span
    }

    fn lex_number(&mut self, c: char) {
        let mut number = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_digit()) {
            number.push(c);
        }
        let span = self.span();
        self.add_token_with_span(
            TokenKind::Number(number.parse().expect("invalid number")),
            span,
        );
    }

    fn lex_identifier(&mut self, c: char) {
        let mut identifier = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_alphanumeric() || c == '_') {
            identifier.push(c);
        }

        let kind = match identifier.as_str() {
            "push" => TokenKind::KeywordPush,
            "pop" => TokenKind::KeywordPop,
            "inc" => TokenKind::KeywordInc,
            "load" => TokenKind::KeywordLoad,
            "add" => TokenKind::KeywordAdd,
            "sub" => TokenKind::KeywordSub,
            "div" => TokenKind::KeywordDiv,
            "mul" => TokenKind::KeywordMul,
            "eq" => TokenKind::KeywordEq,
            "jne" => TokenKind::KeywordJne,
            "hult" => TokenKind::KeywordHult,
            "printreg" => TokenKind::KeywordPrintReg,
            "and" => TokenKind::KeywordAnd,
            "or" => TokenKind::KeywordOr,
            _ => TokenKind::Identifier(identifier),
        };
        let span = self.span();
        self.add_token_with_span(kind, span);
    }

    fn comment(&mut self) {
        while let Some(_) = self.next_if(|c| c != '\n') {}
        self.span();
    }

    fn lex(mut self) -> Vec<Token> {
        while let Some(c) = self.next() {
            match c {
                '0'..='9' => self.lex_number(c),
                'a'..='z' | 'A'..='Z' | '_' => self.lex_identifier(c),
                ' ' | '\t' => {
                    self.span();
                }
                ';' => self.comment(),
                '=' => self.add_token(TokenKind::Equal),
                '\n' => self.add_token(TokenKind::Delimiter),
                '(' => self.add_token(TokenKind::LeftParen),
                ')' => self.add_token(TokenKind::RightParen),
                '{' => self.add_token(TokenKind::LeftBrace),
                '}' => self.add_token(TokenKind::RightBrace),
                '[' => self.add_token(TokenKind::LeftBracket),
                ']' => self.add_token(TokenKind::RightBracket),
                ',' => self.add_token(TokenKind::Comma),
                '.' => self.add_token(TokenKind::Period),
                ':' => self.add_token(TokenKind::Colon),
                '%' => self.add_token(TokenKind::PercentSign),
                '&' => self.add_token(TokenKind::Ampersand),
                _ => panic!("Unexpected character: {}", c),
            }
        }
        self.tokens
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    pub fn snapshot_lexing(input: &str) -> String {
        let tokens = lex(input);
        let mut tokens = std::collections::VecDeque::from(tokens);
        let mut output = String::new();
        for (row, line) in input.lines().enumerate() {
            output += line;
            output += "\n";
            while let Some(tok) = tokens.pop_front() {
                if tok.span.row_end != row {
                    tokens.push_front(tok);
                    break;
                }
                output += &" ".repeat(tok.span.col_start);
                output += &"^".repeat(tok.span.col_end.saturating_sub(tok.span.col_start));
                output += &format!(" {tok:?}");
                output += "\n"
            }
        }
        output
    }

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_lexing(contents));
                });
            }
        };
    }

    snapshot!(demo, "../samples/demo.asm");
    snapshot!(main, "../samples/main.asm");
    snapshot!(fib, "../samples/fib.asm");
}
