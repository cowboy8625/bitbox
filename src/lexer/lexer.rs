use super::token;
use crate::ast::{
    BBString, Builtin, Colon, Comma, Dot, Equals, Identifier, InvalidToken, LeftBrace, LeftParen,
    Number, Plus, RightBrace, RightParen, Semicolon,
};

type Token = Box<dyn token::Token>;
pub struct Lexer<'a> {
    chars: std::iter::Peekable<std::str::Chars<'a>>,
    span: token::Span,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            chars: input.chars().peekable(),
            span: 0..0,
        }
    }

    fn next(&mut self) -> Option<char> {
        let value = self.chars.next();
        if value.is_some() {
            self.span.end += 1;
        }
        value
    }

    fn next_if(&mut self, predicate: impl Fn(char) -> bool) -> Option<char> {
        match self.chars.peek() {
            Some(value) if predicate(*value) => self.next(),
            _ => None,
        }
    }

    fn spanned(&mut self) -> token::Span {
        let span = self.span.clone();
        self.span = self.span.end..self.span.end;
        span
    }

    fn parse_number(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);

        while let Some(value) =
            self.next_if(|value| value.is_ascii_digit() || ['.', '_'].contains(&value))
        {
            lexeme.push(value);
        }

        token::create::<Number>(lexeme, self.spanned())
    }

    fn parse_identifier(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);
        while let Some(value) = self.next_if(|value| value.is_ascii_alphanumeric() || value == '_')
        {
            lexeme.push(value);
        }

        token::create::<Identifier>(lexeme, self.spanned())
    }

    fn parse_builtin(&mut self) -> Token {
        let mut lexeme = String::from('@');
        while let Some(value) = self.next_if(|value| value.is_ascii_alphanumeric() || value == '_')
        {
            lexeme.push(value);
        }

        token::create::<Builtin>(lexeme, self.spanned())
    }

    fn parse_string(&mut self) -> Token {
        let mut lexeme = String::from('#');
        while let Some(value) = self.next() {
            lexeme.push(value);
            if lexeme.ends_with("\"#") {
                break;
            }
        }
        token::create::<BBString>(lexeme, self.spanned())
    }

    fn parse(&mut self) -> Option<Token> {
        match self.next() {
            Some(value @ '0'..='9') => Some(self.parse_number(value)),
            Some(value) if value.is_ascii_alphabetic() => Some(self.parse_identifier(value)),
            Some(value) if value.is_ascii_whitespace() => {
                self.spanned();
                self.parse()
            }
            Some('#') if self.chars.peek() == Some(&'"') => Some(self.parse_string()),
            Some('@') => Some(self.parse_builtin()),
            Some('+') => Some(token::create::<Plus>('+', self.spanned())),
            Some('(') => Some(token::create::<LeftParen>('(', self.spanned())),
            Some(')') => Some(token::create::<RightParen>(')', self.spanned())),
            Some('{') => Some(token::create::<LeftBrace>('(', self.spanned())),
            Some('}') => Some(token::create::<RightBrace>(')', self.spanned())),
            Some(':') => Some(token::create::<Colon>(':', self.spanned())),
            Some(';') => Some(token::create::<Semicolon>(';', self.spanned())),
            Some(',') => Some(token::create::<Comma>(',', self.spanned())),
            Some('=') => Some(token::create::<Equals>('=', self.spanned())),
            Some('.') => Some(token::create::<Dot>('.', self.spanned())),
            Some(value) => Some(token::create::<InvalidToken>(value, self.span.clone())),
            None => None,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        self.parse()
    }
}

#[test]
fn test_lexer() {
    let token = Lexer::new("#\"test\"#").parse().unwrap();
    assert_eq!(token.get_lexeme(), "#\"test\"#");
}
