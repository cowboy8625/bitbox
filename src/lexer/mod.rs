mod lexer;
pub mod token;

#[cfg(test)]
mod test;

pub fn lex(src: &str) -> Vec<token::Token> {
    lexer::Lexer::new(src).collect()
}
