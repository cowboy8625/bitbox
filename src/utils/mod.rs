mod either;
pub use either::Either;
