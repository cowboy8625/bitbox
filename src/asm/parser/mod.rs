use crate::asm::lexer::{Span, Token, TokenKind};
use crate::instructions::{Data, Imm, Instruction, Label, Opcode, Register, Type};
use crate::utils::Either;
use anyhow::{bail, Result};
use std::iter::Peekable;
use thiserror::Error;

pub fn parse(src: &str) -> Result<Vec<Item>> {
    let tokens = super::lexer::lex(src);
    let parser = Parser::new(tokens);
    parser.parse()
}

#[derive(Debug, Error)]
struct BatchParserErrors(Vec<anyhow::Error>);

impl std::fmt::Display for BatchParserErrors {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for e in &self.0 {
            writeln!(f, "{}", e)?;
        }
        Ok(())
    }
}

#[derive(Debug, Error)]
pub enum ParserError {
    #[error("Unknown directive {0:?} at {}..{}", .1.col_start, .1.col_end)]
    UnknownDirective(String, Span),
    #[error("Missing entry point")]
    MissingEntryPoint,
    #[error("Expected identifier at {}..{} but found {0:?}", .1.col_start, .1.col_end)]
    ExpectedIdentifier(TokenKind, Span),
    #[error("Expected number at {}..{} but found {0:?}", .1.col_start, .1.col_end)]
    ExpectedNumber(TokenKind, Span),
    #[error("Expected delimiter at {}..{} but found {0:?}", .1.col_start, .1.col_end)]
    ExpectedDelimiter(TokenKind, Span),
    #[error("Expected percent sign at {}..{} but found {0:?}", .1.col_start, .1.col_end)]
    ExpectedPercentSign(TokenKind, Span),
    #[error("Expected colon at {}..{} but found {0:?}", .1.col_start, .1.col_end)]
    ExpectedColon(TokenKind, Span),
    #[error("Unexpected EOF")]
    UnexpectedEof,
    #[error("Expected Right Bracket but found {0:?} at {1:?}")]
    ExpectedRightBracket(TokenKind, Span),
    #[error("Expected Left Bracket but found {0:?} at {1:?}")]
    ExpectedLeftBracket(TokenKind, Span),
    #[error("Expected Signed or Unsigned but found {0:?} at {1:?}")]
    ExpectedSign(TokenKind, Span),
    #[error("Invalid Imm Type found {0:?} at {1:?}")]
    InvalidImmType(TokenKind, Span),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Item {
    EntryPoint(Token),
    // Data(Vec<Data>),
    Text(Vec<Text>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Text {
    pub label: Option<Label>,
    pub opcode: Instruction,
    pub span: Span,
}

impl Text {
    pub fn new_opcode(opcode: Instruction, span: Span) -> Self {
        Self {
            label: None,
            opcode,
            span,
        }
    }
    pub fn new_opcode_with_label(label_token: Token, opcode: Instruction, span: Span) -> Self {
        let TokenKind::Identifier(name) = label_token.kind else {
            panic!("Expected identifier")
        };
        Self {
            label: Some(Label {
                name,
                span: label_token.span,
                def: true,
            }),
            opcode,
            span,
        }
    }

    pub fn _span(&self) -> Span {
        let Some(label) = &self.label else {
            return self.span;
        };
        Span::from((label.span, self.span))
    }
}

struct Parser {
    tokens: Peekable<std::vec::IntoIter<Token>>,
    instructions: Vec<Item>,
    text_section: Vec<Text>,
    // data_section: Vec<Data>,
    entry_point: Option<Token>,
    label: Option<Token>,
    errors: Vec<anyhow::Error>,
}

impl Parser {
    fn new(tokens: Vec<Token>) -> Self {
        Self {
            tokens: tokens.into_iter().peekable(),
            instructions: Vec::new(),
            text_section: Vec::new(),
            // data_section: Vec::new(),
            entry_point: None,
            label: None,
            errors: Vec::new(),
        }
    }

    fn next(&mut self) -> Option<Token> {
        self.tokens.next()
    }

    fn peek(&mut self) -> Option<&Token> {
        self.tokens.peek()
    }

    fn peek_kind(&mut self) -> Option<&TokenKind> {
        self.peek().map(|t| &t.kind)
    }

    fn push_text(&mut self, text: Text) {
        self.text_section.push(text)
    }

    fn create_text(&mut self, opcode: Instruction, span: Span) -> Text {
        let Some(label) = self.label.take() else {
            return Text::new_opcode(opcode, span);
        };
        Text::new_opcode_with_label(label, opcode, span)
    }

    fn consume(
        &mut self,
        kinds: &[TokenKind],
        error: impl Fn(TokenKind, Span) -> ParserError,
    ) -> Result<Token> {
        let Some(token) = self.next() else {
            bail!(ParserError::UnexpectedEof);
        };
        if !kinds.contains(&token.kind) {
            bail!(error(token.kind, token.span));
        }
        Ok(token)
    }

    fn parse_reg(&mut self) -> Result<Register> {
        let _ = self.consume(&[TokenKind::PercentSign], ParserError::ExpectedPercentSign)?;
        let Some(token) = self.next() else {
            bail!(ParserError::UnexpectedEof);
        };
        let TokenKind::Number(reg) = token.kind else {
            bail!(ParserError::ExpectedNumber(token.kind, token.span));
        };

        Ok(Register::try_from((reg as u8, token.span))?)
    }

    fn parse_type(&mut self) -> Result<Type> {
        let _ = self.consume(&[TokenKind::LeftBracket], ParserError::ExpectedLeftBracket)?;
        let Some(sign_token) = self.next() else {
            bail!(ParserError::UnexpectedEof);
        };
        let TokenKind::Identifier(ref t) = sign_token.kind else {
            bail!(ParserError::ExpectedIdentifier(
                sign_token.kind,
                sign_token.span
            ));
        };

        let sign = &t[0..1];
        let bits = &t[1..].parse::<u8>()?;

        let _ = self.consume(
            &[TokenKind::RightBracket],
            ParserError::ExpectedRightBracket,
        )?;
        let r#type = match sign.to_lowercase().as_str() {
            "i" => Type::I(*bits),
            "u" => Type::U(*bits),
            _ => bail!(ParserError::ExpectedSign(sign_token.kind, sign_token.span)),
        };
        Ok(r#type)
    }

    fn parse_value(&mut self, t: Type) -> Result<Vec<u8>> {
        let token = match self.next() {
            Some(token) => token,
            None => bail!(ParserError::UnexpectedEof),
        };

        let value = match token.kind {
            TokenKind::Number(value) => value,
            _ => bail!(ParserError::ExpectedNumber(token.kind, token.span)),
        };

        match t {
            Type::U(8) | Type::I(8) => Ok(vec![value as u8]),
            Type::U(16) | Type::I(16) => Ok(vec![value as u8, (value >> 8) as u8]),
            Type::U(32) | Type::I(32) => Ok(vec![
                value as u8,
                (value >> 8) as u8,
                (value >> 16) as u8,
                (value >> 24) as u8,
            ]),
            Type::U(64) | Type::I(64) => Ok(vec![
                value as u8,
                (value >> 8) as u8,
                (value >> 16) as u8,
                (value >> 24) as u8,
                // TODO: This can never be 64 bits unless we can change the TokenKind::Number to
                // hold a u64 or more
                // (value >> 32) as u8,
                // (value >> 40) as u8,
                // (value >> 48) as u8,
                // (value >> 56) as u8,
            ]),
            Type::Void => bail!(ParserError::InvalidImmType(token.kind, token.span)),
            _ => unimplemented!("{:?} is not implemented", t),
        }
    }

    fn parse_reg_imm(&mut self, token: Token, opcode: Opcode) -> Result<()> {
        let r#type = self.parse_type()?;
        let reg = self.parse_reg()?;
        let value = self.parse_value(r#type)?;
        let end_span = self
            .consume(&[TokenKind::Delimiter], ParserError::ExpectedDelimiter)?
            .span;
        let span = Span::from((token.span, end_span));
        let instruction = Instruction {
            opcode,
            r#type,
            data: Data::Imm(reg, Imm(value)),
        };
        let text = self.create_text(instruction, span);
        self.push_text(text);
        Ok(())
    }

    fn _parse_reg_2(&mut self, token: Token, opcode: Opcode) -> Result<()> {
        let r#type = self.parse_type()?;
        let des = self.parse_reg()?;
        let lhs = self.parse_reg()?;
        let end_span = self
            .consume(&[TokenKind::Delimiter], ParserError::ExpectedDelimiter)?
            .span;
        let span = Span::from((token.span, end_span));
        let instruction = Instruction {
            opcode,
            r#type,
            data: Data::Reg2(des, lhs),
        };
        let text = self.create_text(instruction, span);
        self.push_text(text);
        Ok(())
    }

    fn parse_reg_3(&mut self, token: Token, opcode: Opcode) -> Result<()> {
        let r#type = self.parse_type()?;
        let des = self.parse_reg()?;
        let lhs = self.parse_reg()?;
        let rhs = self.parse_reg()?;
        let end_span = self
            .consume(&[TokenKind::Delimiter], ParserError::ExpectedDelimiter)?
            .span;

        let span = Span::from((token.span, end_span));
        let instruction = Instruction {
            opcode,
            r#type,
            data: Data::Reg3(des, lhs, rhs),
        };
        let text = self.create_text(instruction, span);
        self.push_text(text);
        Ok(())
    }

    fn parse_reg_1(&mut self, token: Token, opcode: Opcode) -> Result<()> {
        let r#type = self.parse_type()?;
        let reg = self.parse_reg()?;
        let end_span = self
            .consume(&[TokenKind::Delimiter], ParserError::ExpectedDelimiter)?
            .span;
        let span = Span::from((token.span, end_span));
        let instruction = Instruction {
            opcode,
            r#type,
            data: Data::Reg1(reg),
        };
        let text = self.create_text(instruction, span);
        self.push_text(text);
        Ok(())
    }

    fn parse_no_args(&mut self, token: Token, opcode: Opcode) -> Result<()> {
        let end_span = self
            .consume(&[TokenKind::Delimiter], ParserError::ExpectedDelimiter)?
            .span;
        let span = Span::from((token.span, end_span));
        let instruction = Instruction {
            opcode,
            r#type: Type::Void,
            data: Data::NoArgs,
        };
        let text = self.create_text(instruction, span);
        self.push_text(text);
        Ok(())
    }

    fn parse_reg_2_label(&mut self, token: Token, opcode: Opcode) -> Result<()> {
        let r#type = Type::Void;
        let lhs = self.parse_reg()?;
        let rhs = self.parse_reg()?;
        let label = self.parse_label()?;
        let end_span = self
            .consume(&[TokenKind::Delimiter], ParserError::ExpectedDelimiter)?
            .span;
        let span = Span::from((token.span, end_span));
        let instruction = Instruction {
            opcode,
            r#type,
            data: Data::RegLabel(lhs, rhs, Either::Left(label)),
        };
        let text = self.create_text(instruction, span);
        self.push_text(text);

        Ok(())
    }

    fn entry_point(&mut self) -> Result<()> {
        // TODO: Make span match full entry point
        let token = self.next();

        if token
            .clone()
            .map(|t| !matches!(t.kind, TokenKind::Identifier(_)))
            .unwrap_or(true)
        {
            bail!(ParserError::ExpectedIdentifier(
                token.clone().unwrap().kind,
                token.unwrap().span
            ));
        }
        self.entry_point = token;
        let _ = self.consume(&[TokenKind::Delimiter], ParserError::ExpectedDelimiter)?;
        Ok(())
    }

    fn parse_label(&mut self) -> Result<Label> {
        let Some(token) = self.next() else {
            bail!(ParserError::UnexpectedEof);
        };
        let TokenKind::Identifier(name) = token.kind else {
            bail!(ParserError::ExpectedIdentifier(token.kind, token.span));
        };
        Ok(Label {
            name,
            span: token.span,
            def: false,
        })
    }

    fn parse_directive(&mut self, token: Token) -> Result<()> {
        let identifier = self.next();
        let Some(Token { kind, span }) = identifier else {
            bail!(ParserError::ExpectedIdentifier(token.kind, token.span));
        };
        let TokenKind::Identifier(name) = kind else {
            bail!(ParserError::ExpectedIdentifier(kind, span));
        };
        match name.as_str() {
            "entry" => self.entry_point()?,
            _ => bail!(ParserError::UnknownDirective(name, span)),
        }
        Ok(())
    }

    fn parse(mut self) -> Result<Vec<Item>> {
        while let Some(token) = self.next() {
            let kind = token.kind.clone();
            let maybe_error = match kind {
                TokenKind::Period => self.parse_directive(token),
                TokenKind::KeywordPush => self.parse_reg_1(token, Opcode::Push),
                TokenKind::KeywordPop => self.parse_reg_1(token, Opcode::Pop),
                TokenKind::KeywordLoad => self.parse_reg_imm(token, Opcode::Load),
                TokenKind::KeywordAdd => self.parse_reg_3(token, Opcode::Add),
                TokenKind::KeywordSub => self.parse_reg_3(token, Opcode::Sub),
                TokenKind::KeywordDiv => self.parse_reg_3(token, Opcode::Div),
                TokenKind::KeywordMul => self.parse_reg_3(token, Opcode::Mul),
                TokenKind::KeywordEq => self.parse_reg_3(token, Opcode::Eq),
                TokenKind::KeywordInc => self.parse_reg_1(token, Opcode::Inc),
                TokenKind::KeywordJne => self.parse_reg_2_label(token, Opcode::Jne),
                TokenKind::KeywordHult => self.parse_no_args(token, Opcode::Hult),
                TokenKind::KeywordPrintReg => self.parse_reg_1(token, Opcode::PrintReg),
                TokenKind::KeywordAnd => self.parse_reg_3(token, Opcode::And),
                TokenKind::KeywordOr => self.parse_reg_3(token, Opcode::Or),
                TokenKind::Identifier(_) if matches!(self.peek_kind(), Some(&TokenKind::Colon)) => {
                    self.label = Some(token);
                    let _ = self.consume(&[TokenKind::Colon], ParserError::ExpectedColon)?;
                    continue;
                }
                TokenKind::Delimiter => continue,
                _ => todo!("report error {:?}", kind),
            };
            let Err(e) = maybe_error else {
                continue;
            };
            self.errors.push(e);
        }
        if self.entry_point.is_none() {
            self.errors.push(ParserError::MissingEntryPoint.into());
        }
        if !self.errors.is_empty() {
            bail!(BatchParserErrors(self.errors));
        }
        self.instructions
            .push(Item::EntryPoint(self.entry_point.unwrap()));
        self.instructions.push(Item::Text(self.text_section));
        Ok(self.instructions)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snapshot_entry_point(line: &str, entry_item: Option<Item>) -> Result<String> {
        let mut output = String::new();
        let Some(Item::EntryPoint(Token { span, kind })) = entry_item else {
            panic!("Expected entry point but found: {:?}", entry_item);
        };
        output += line;
        output += "\n";
        output += &" ".repeat(span.col_start);
        output += &"^".repeat(span.col_end - span.col_start);
        output += &format!(" {kind:?}");
        output += "\n";
        Ok(output)
    }

    pub fn snapshot_parsing(input: &str) -> Result<String> {
        let items = parse(input)?;
        let mut line_iter = input.lines();
        let Some(line) = line_iter.next() else {
            panic!("Expected at least one line");
        };

        let mut items = std::collections::VecDeque::from(items);
        let mut output = snapshot_entry_point(line, items.pop_front())?;
        let text_section = items.pop_front();
        let Some(Item::Text(text_section)) = text_section else {
            panic!("Expected text section but found: {:?}", text_section);
        };

        let mut text_section = std::collections::VecDeque::from(text_section);
        for (row, line) in line_iter.enumerate() {
            output += line;
            output += "\n";
            while let Some(text) = text_section.pop_front() {
                if text.span.row_end != row {
                    text_section.push_front(text);
                    break;
                }
                output += &" ".repeat(text.span.col_start);
                output += &"^".repeat(text.span.col_end - text.span.col_start);
                output += &format!(" {text:?}");
                output += "\n"
            }
        }
        Ok(output)
    }

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_parsing(contents).unwrap());
                });
            }
        };
    }

    snapshot!(demo, "../samples/demo.asm");
    snapshot!(main, "../samples/main.asm");
}
