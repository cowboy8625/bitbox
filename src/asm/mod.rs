mod lexer;
mod parser;
use crate::asm::{
    lexer::{Token, TokenKind},
    parser::Item,
};
use std::collections::HashMap;

use anyhow::Result;

use self::parser::Label;

type SymbolTable = HashMap<String, u32>;

pub use lexer::Span;

pub fn assemble(src: &str) -> Result<Vec<u8>> {
    let program_instructions = parser::parse(src)?;
    let data_offset = 0;
    let symbol_table = symbol_table(&program_instructions, data_offset);
    let mut program = Vec::new();
    let mut entry_point: Option<u32> = None;
    for item in program_instructions.iter() {
        match item {
            Item::EntryPoint(token) => {
                let Token {
                    kind: TokenKind::Identifier(name),
                    ..
                } = token
                else {
                    unreachable!("expected identifier")
                };
                let index = &0; // symbol_table.get(name).unwrap();
                entry_point = Some(*index + Header::SIZE as u32);
            }
            Item::Text(text) => {
                for instruction in text.iter() {
                    let opcode = instruction.opcode.to_bytes();
                    program.extend_from_slice(&opcode);
                }
            }
        }
    }

    let mut header = Header::default();
    // TODO: set the length of the data section here
    header.set_header_text_section(0);
    let Some(entry_point) = entry_point else {
        return Err(anyhow::anyhow!("no entry point found"));
    };
    header.set_header_entry_point(entry_point);
    let mut bin = header.build().to_vec();
    bin.extend_from_slice(&program);
    Ok(bin)
}

#[derive(Debug)]
pub struct Header([u8; Header::SIZE]);
impl Header {
    pub const MAGIC_NUMBER: [u8; 4] = [0x42, 0x42, 0x56, 0x4d]; // BBVM
    pub const TEXT_OFFSET: usize = 4;
    pub const ENTRY_OFFSET: usize = 8;
    pub const SIZE: usize = 64;

    /// Takes the length of bytes of the data section
    fn set_header_text_section(&mut self, offset: u32) {
        let [a, b, c, d] = offset.to_le_bytes();
        self.0[Self::TEXT_OFFSET + 0] = a;
        self.0[Self::TEXT_OFFSET + 1] = b;
        self.0[Self::TEXT_OFFSET + 2] = c;
        self.0[Self::TEXT_OFFSET + 3] = d;
    }

    fn set_header_entry_point(&mut self, offset: u32) {
        let [a, b, c, d] = offset.to_le_bytes();
        self.0[Self::ENTRY_OFFSET + 0] = a;
        self.0[Self::ENTRY_OFFSET + 1] = b;
        self.0[Self::ENTRY_OFFSET + 2] = c;
        self.0[Self::ENTRY_OFFSET + 3] = d;
    }

    fn build(self) -> [u8; Header::SIZE] {
        self.0
    }
}

impl Default for Header {
    fn default() -> Self {
        let mut bytes = [0; Header::SIZE];
        bytes[0] = Header::MAGIC_NUMBER[0];
        bytes[1] = Header::MAGIC_NUMBER[1];
        bytes[2] = Header::MAGIC_NUMBER[2];
        bytes[3] = Header::MAGIC_NUMBER[3];
        Self(bytes)
    }
}

fn symbol_table(instructions: &[Item], data_offset: u32) -> SymbolTable {
    let mut symbol_table = SymbolTable::new();
    let mut ip = data_offset + Header::SIZE as u32;
    for item in instructions.iter() {
        match item {
            Item::EntryPoint(_) => {}
            Item::Text(text) => {
                for instruction in text.iter() {
                    if let Some(Label {
                        name, def: true, ..
                    }) = &instruction.label
                    {
                        symbol_table.insert(name.clone(), ip);
                        ip += 4;
                    } else {
                        ip += 4;
                    }
                }
            }
        }
    }
    symbol_table
}
