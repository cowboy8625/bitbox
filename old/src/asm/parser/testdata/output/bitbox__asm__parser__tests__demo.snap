---
source: src/asm/parser/mod.rs
expression: snapshot_parsing(contents).unwrap()
---
.entry main
       ^^^^ Identifier("main")

load[u32] %0 123
load[u32] %1 321
^^^^^^^^^^^^^^^^^ Text { label: None, opcode: Instruction { opcode: Load, type: U(32), data: Imm(R0, Imm([123, 0, 0, 0])) }, span: Span { row_start: 1, row_end: 2, col_start: 0, col_end: 17, byte_start: 13, byte_end: 30 } }
add[u32] %0 %0 %1
^^^^^^^^^^^^^^^^^ Text { label: None, opcode: Instruction { opcode: Load, type: U(32), data: Imm(R1, Imm([65, 1, 0, 0])) }, span: Span { row_start: 2, row_end: 3, col_start: 0, col_end: 17, byte_start: 30, byte_end: 47 } }
hult
^^^^^^^^^^^^^^^^^^ Text { label: None, opcode: Instruction { opcode: Add, type: U(32), data: Reg3(R0, R0, R1) }, span: Span { row_start: 3, row_end: 4, col_start: 0, col_end: 18, byte_start: 47, byte_end: 65 } }
